1-theory/3-api-contracts.adoc

```diff
@@ -0,0 +1,445 @@
+`components.schema` is describing data model. The schema keywords that OpenAPI
+v3.0 uses are based on JSON Schema, and there is a bit of a tangent we should
+look into here about compatibility.
+
+[[WARNING]]
+====
+OpenAPI v3.0 schema objects are a subset/superset/sideset implementation of
+_JSON Schema draft 05_. Most JSON Schema keywords as expected, a few extra
+OpenAPI-only keywords were added, and some JSON Schema keywords are not
+supported. There is also the tricky situation where JSON Schema has continued to
+progress quickly since draft 4 (draft 8 is almost complete at time of writing).
+
+This does cause a bit of confusion for new developers, but interoperability
+amongst standards is always a tricky one. Thankfully, future versions of OpenAPI
+(probably v3.1) aim to solve this, so we won't get too stuck in the weeds here.
+For those who want to learn more,
+https://blog.apisyouwonthate.com/openapi-and-json-schema-divergence-part-1-1daf6678d86e[this
+first article] fully explains the situation, and
+https://blog.apisyouwonthate.com/openapi-and-json-schema-divergence-part-2-52e282e06a05[this
+second article] explains workarounds and longer term solutions.
+====
+
+OpenAPI contract files are usually static. They're usually written down along
+with the source code, then sometimes deployed to a file hosting service like S3
+for folks to use. Some managers want to treat these like business secrets and
+hide them under lock and key, which makes absolutely no damned sense as they are
+meta-data only. Most "hackers" could probably figure out that you keep your list
+of companies under `GET /companies`, so just don't make that a publicly
+available endpoint and you're gonna be ok. PayPal, Microsoft, and other
+companies make their OpenAPI contracts available to anyone who wants to download
+them, and this approach can help folks integrate with your APIs.
+
+You can imagine an OpenAPI file growing to be rather unwieldy once its got 50+
+endpoints and more complex examples, but have no fear you can spread things
+around in multiple files to make it a lot more DRY (Don't Repeat Yourself) and
+useful. The first thing to go is usually the `components.schema` definitions,
+which can be moved to their own files. Once these are split into their own
+files, an extra step can be taken to turn them into proper JSON Schema files.
+Once they are split out they can be referenced in a HTTP response header.
+
+[source]
+----
+Link: <http://example.com/schemas/hat.json#>; rel="describedby"
+----
+
+When a client sees this they can use it for all sorts of things - like form
+generation and client-side validation - all without needing to figure out how to
+distribute the files to them ahead of time.
+
+==== GraphQL Schemas
+
+GraphQL as an implementation comes bundled with
+https://graphql.org/learn/schema/[GraphQL Schemas]. GraphQL does not really have
+a service model, as it does not need one. All interactions operate under a
```

Comment:
Fair points! I think I made the "it can be anywhere including AMQP but in HTTP land its usually just a POST endpoint" in the paradigms & implementations chapter. :)

Reply:
Ah no found a way to improve it. Great feedback.